src/oscar/apps/address/abstract_models.py

```diff
@@ -351,6 +344,36 @@ def join_fields(self, fields, separator=u", "):
             else:
                 value = getattr(self, field)
             field_values.append(value)
+        return field_values
+
+    def get_address_field_values(self, fields, include_salutation=True, include_country=True):
+        """
+        Returns set of field values within the salutation and country.
+        """
+        field_values = self.get_field_values(fields)
+        if include_salutation:
+            field_values = [self.salutation] + self.get_field_values(fields)
+        field_values = [f.strip() for f in field_values if f]
+        if include_country:
+            try:
+                field_values.append(self.country.printable_name)
+            except exceptions.ObjectDoesNotExist:
+                pass
+        return field_values
+
+    def generate_hash(self):
+        """
+        Returns a hash of the address, based on standard set of fields, listed
+        out in `hash_fields` property.
+        """
+        field_values = self.get_address_field_values(self.hash_fields)
+        return zlib.crc32(', '.join(field_values).upper().encode('UTF8')) & 0xffffffff
```

Comment:
What is the `0xffffffff` for here? Also, is the hash that this generates backward compatible (i.e., will it be the same hash that the previous method generated)?

Reply:
> What is the 0xffffffff for here?

This expression allows to generate the same CRC value for both Python 2 and Python 3 version, since by default they have different range of values.

> Also, is the hash that this generates backward compatible (i.e., will it be the same hash that the previous method generated)?

It appears, that for Python 2 - no, since previously we did not enforce it to return same value for all Python versions.

Reply:
I think it would be problematic to break backward compatibility since the hash is stored in the DB and used for comparisons. We either need to ensure that it stays the same, or add a migration to update all existing hashes (which could be very slow if there are a large number)?

Reply:
I think firstly we need to define if we want hash to be generate the same for all versions. I believe that it's necessary and in the future ubiquitous upgrade to Python 3 is inevitable so sooner or later it has to be done.

Reply:
I think the hash has to be consistent given what we use it for. Maybe we need a migration that converts the old Python 2 hashes into the new ones.